botbot_plugins/plugins/vote.py

```diff
@@ -0,0 +1,334 @@
+    "invalid_option": u"“{option}” is not a valid option."
+}
+INFO_MESSAGES = {
+    "voting_started": u"Voting has started.",
+    "voting_started_name": u"Voting has started for proposal “{name}”.",
+    "voting_ended": u"Voting has ended.",
+    "voting_ended_name": u"Voting has ended for proposal “{name}”."
+}
+
+
+def get_section_template(name, for_, against, username_sep):
+    template = (u"[{}(+{}, -{}): "
+                .format(name, len(for_), len(against)))
+
+    if for_:
+        template += username_sep.join(for_)
+    if for_ and against:
+        template += u"; "
+    if against:
+        template += username_sep.join(
+            [u"-" + username for username in against])
+    template += u"]"
+    return template
+
+
+COUNTVOTE_TEMPLATES = {
+    "default_for_template": u"[+{number_for}: {users}]",
+    "default_against_template": u"[-{number_against}: {users}]",
+    "abstain_template": u"[\\{number_abstain}: {users}]",
+    "section_template": get_section_template
+}
+
+FOR_NAME = "for"
+AGAINST_NAME = "against"
+ABSTAIN_NAME = "abstain"
+OPPOSING_VOTE = {
+    FOR_NAME: AGAINST_NAME,
+    AGAINST_NAME: FOR_NAME
+}
+VOTE_SHORTCUT_TO_NAME = {
+    "": FOR_NAME,
+    "+": FOR_NAME,
+    "-": AGAINST_NAME
+}
+
+Vote = namedtuple('Vote', ['voter', 'section', 'vote'])
+
+class Config(config.BaseConfig):
+
+    options_separator = config.Field(help_text="Separator to use when starting a vote with custom options", default=",")
+    username_separator = config.Field(help_text="Separator between usernames when printing out votes", default=",")
+    default_vote_option = config.Field(help_text="The key under which to store when no options are given", default="__vote__")
+    abstain_option = config.Field(help_text="The key under which to store abstaining voters", default="__abstain__")
+    boolean_shortform = config.Field(help_text=u"The short form that voters use when they’re boolean voting", default="1")
+
+class Plugin(BasePlugin):
+    """
+    Proposal voting system
+    """
+
+    config_class = Config
+
+    @listens_to_regex_command("startvote",
+                              ur"(?P<name>[^\[\]]+?)?\s*(\[(?P<options>.*)\])?\s*$")
+    def startvote(self, line, name, options):
+        options_sep = self.config["options_separator"]
+        if self.retrieve("votes"):
+            return (ERROR_MESSAGES["voting_already_running"].
+                    format(author=line.user))
+
+        default_options = [self.config["default_vote_option"]]
+        options = options.split(options_sep) if options else []
+        options = [option_name.strip() for option_name in options
+                   if not option_name.isspace()]
+        options = options if options else default_options
+        # Abstaining should always be an option
+        options.append(self.config["abstain_option"])
+        options = list(set(options))
```

Comment:
Related to my previous comment, not reassigning `options` a bunch of times might help.

Reply:
Reduced the number of times this happens but haven't made it so that it's assigned in one go.